src/sys/pthread.rs

```diff
@@ -34,10 +39,351 @@ pub fn pthread_kill<T>(thread: Pthread, signal: T) -> Result<()>
+///     ForkResult::Parent { child } => {
+///         let guard = mutex.lock()?;
+///         sleep(TIMEOUT);
+///         guard.try_unlock();
+///         // Wait for child process to exit
+///         unsafe {
+///             assert_eq!(libc::waitpid(child.as_raw(), std::ptr::null_mut(),0), child.as_raw());
+///         }
+///     },
+///     ForkResult::Child => {
+///         let now = Instant::now();
+///         sleep(DELTA);
+///         mutex.lock()?;
+///         assert!(now.elapsed() > TIMEOUT && now.elapsed() < TIMEOUT + DELTA);
+///     }
+/// }
+/// 
+/// # Ok(())
+/// # }
+/// ```
+#[cfg(target_os = "linux")]
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct Mutex(UnsafeCell<libc::pthread_mutex_t>);
+
+#[cfg(target_os = "linux")]
+impl Mutex {
+    /// Wraps [`libc::pthread_mutex_init`].
+    /// 
+    /// # Safety
+    /// 
+    /// Attempting to initialize an already initialized mutex results in undefined behavior.
+    pub unsafe fn init(mutex: *mut Mutex, attr: Option<MutexAttr>) -> Result<()> {
+        let attr = match attr {
+            Some(mut x) => &mut x.0,
+            None => std::ptr::null_mut()
+        };
+        Errno::result(libc::pthread_mutex_init((*mutex).0.get(),attr))?;
+        Ok(())
+    }
+    /// Wraps [`libc::pthread_mutex_init`].
+    pub fn new(attr: Option<MutexAttr>) -> Result<Self> {
+        let attr = match attr {
+            Some(mut x) => &mut x.0,
+            None => std::ptr::null_mut()
+        };
+        let init = unsafe {
+            let mut uninit = std::mem::MaybeUninit::<libc::pthread_mutex_t>::uninit();
+            Errno::result(libc::pthread_mutex_init(uninit.as_mut_ptr(),attr))?;
+            uninit.assume_init()
+        };
+        Ok(Self(UnsafeCell::new(init)))
+    }
+    /// Wraps [`libc::pthread_mutex_lock`].
+    /// 
+    /// <https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html>
+    pub fn lock(&self) -> Result<MutexGuard<'_>> {
+        unsafe {
+            Errno::result(libc::pthread_mutex_lock(self.0.get())).map(|_| MutexGuard(self))
+        }
+    }
+    /// Wraps [`libc::pthread_mutex_trylock`].
+    /// 
+    /// <https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html>
+    pub fn try_lock(&self) -> Result<Option<MutexGuard<'_>>> {
+        unsafe {
+            match Errno::result(libc::pthread_mutex_trylock(self.0.get())) {
+                Ok(_) => Ok(Some(MutexGuard(self))),
+                Err(Errno::EBUSY) => Ok(None),
+                Err(err) => Err(err)
+            }
+            
+        }
+    }
+    /// Wraps [`libc::pthread_mutex_unlock`].
+    /// 
+    /// <https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html>
+    /// 
+    /// Prefer unlocking by dropping the [`MutexGuard`] returned by [`Mutex::lock`] or [`Mutex::try_lock`].
+    /// 
+    /// # Safety
+    /// 
+    /// Results in UB if not called from the thread that locked the mutex.
+    pub unsafe fn unlock(&self) -> Result<()> {
```

Comment:
Can this be a private method?  Why would a user ever want to call it?

Reply:
See https://github.com/nix-rust/nix/pull/1950#discussion_r1342199592